test/parallel/test_process_sets_static.py

```diff
@@ -0,0 +1,53 @@
+import os
+import sys
+import unittest
+
+import horovod.tensorflow as hvd
+
+sys.path.append(os.path.join(os.path.dirname(__file__), os.pardir, 'utils'))
+
+from common import mpi_env_rank_and_size
+
+class ProcessSetsStaticTests(unittest.TestCase):
+    """ Since this test case initializes Horovod and shuts it down, it must be run in a separate process. """
+    def test_static(self):
+        mpi_rank, mpi_size = mpi_env_rank_and_size()
+        gloo_size = int(os.getenv('HOROVOD_SIZE', -1))
+        if gloo_size != -1:
+            self.skipTest("Multiple process sets currently do not support Gloo controller.")
+
+        size = max(mpi_size, gloo_size)
+
+        # This test does not apply if there is only one worker.
+        if size == 1:
+            self.skipTest("Only one worker available")
+
+        try:
+            import mpi4py
+            mpi4py.rc.initialize = False
+        except ImportError:
+            self.skipTest("This test requires mpi4py.")
+
+        if mpi_rank == 0:
+            hvd.init(process_sets=[hvd.ProcessSet([0]),
```

Comment:
I think this test is a good place to discuss where I think the ProcessSet API can be improved a bit. 
I think we should treat the `ProcessSet` objects as somewhat analogous to a distributed group in the PyTorch distributed API (https://pytorch.org/docs/stable/distributed.html#torch.distributed.new_group) where the specific `ProcessSet` object that is either passed into `hvd.init` or returned from `hvd.add_process_set` acts as the handle for later Horovod operations on the process group. 
As an example, I think a call like this should be designed to be incorrect usage:
```
hvd.init(process_sets=[hvd.ProcessSet([2,3])])
```
as a temporary `ProcessSet` object is passed into init and after the call, the user no longer can access the handle (or should not be able to).

I think correct usage should look something like this:
```
ps = hvd.ProcessSet([2,3]) # Created but not initialized. Usage in any collective op will raise an error
hvd.init(process_sets=[ps]) # Initializes Horovod and sets up passed in ProcessSet ps
hvd.allreduce(tensor, process_set=ps)
```
This also then removes the need to have an API to extract existing process sets from Horovod (along with process set IDs). Essentially, it would be the responsibility of the user to maintain ownership of the ProcessSet objects if they intend to use them.

What do you think?

Reply:
Hey @romerojosh,

thanks for the second review and making sure we will end up with a really well rounded interface! :)

I kind of liked having both options, i.e., letting users either use the `ProcessSet` objects or those IDs (mainly because in the training code we run on this branch we convert tensor or variable names with encoded IDs into process set arguments all the time). _However_, having given this some more thought, I think you are right that this exposed implementation detail will probably feel more confusing than useful to users. They can just have their own `ProcessSet` object storages that will generally be better adapted to their actual needs.

I will give this a second pass of clean ups. Maybe we can completely get rid of the Horovod-internal dict `_id_to_process_sets` as well.

Reply:
OK, things should be a bit simpler now. 

The biggest caveat I can think of is that duplicate process sets aren't handled transparently anymore. So if one tries to have `[ProcessSet(range(0, size, 2)), ProcessSet([0])]` this would now raise an exception when running with two or three processes, but be fine with four or more processes. But people should take a moment to think about their process set setups anyway.

(Allowing these duplicates without Horovod-internal object tracking is dangerous in combination with `remove_process_set()`)